### PR TITLE
add phgarfield drift to laser cluster helper

### PR DESCRIPTION
**Pull request by devloom**
[comment]: <> (Please tell us something about this pull request)
Adds getClusterCentroidWithPHGarfield to the LaserClusterHelper which takes laser cluster centroids from readout and reverse drifts them along polylines to the central membrane.

## Types of changes
[comment]: <> ( What types of changes does your code introduce? Put an `x` in all the boxes that apply: )
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work for users)
- [ ] Requiring change in macros repository (Please provide links to the macros pull request in the last section)
- [x] I am a member of [GitHub organization of sPHENIX Collaboration](https://github.com/orgs/sPHENIX-Collaboration/people), EIC, or ECCE (contact Chris Pinkenburg to join)

## What kind of change does this PR introduce? (Bug fix, feature, ...)

[comment]: <> ( What does this PR do? Linking to talk in software meeting encouraged )

## TODOs (if applicable)

[comment]: <> ( In case this is a draft PR, e.g. for running checks using Jenkins, please make the pull request as a draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/  )

## Motivation

Enable reverse-drift of laser cluster centroids from the readout plane to the central membrane with PHGarfield. Update the build order so PHGarfield is available before the TPC package.

## Key changes

- Add `LaserClusterHelper::getClusterCentroidWithPHGarfield`.
- Configure PHGarfield from CDB field-map and `kEff` calibration data.
- Reverse-drift valid centroids in global and local coordinate systems.
- Add explicit reverse-drift termination statuses.
- Manage PHGarfield with `std::unique_ptr`.
- Add Garfield configuration setters.
- Link the TPC library with PHGarfield.
- Correct the build order for Jenkins builds.

## Potential risk areas

- Reverse-drift can change reconstruction results.
- Incorrect geometry, coordinate, field-map, or calibration settings can produce invalid centroids.
- Non-central-membrane termination statuses require careful handling.
- Reverse-drift can increase processing time.
- PHGarfield ownership and configuration require thread-safety review.
- No IO format changes are described.

## Possible future improvements

- Add tests for coordinate conversion, invalid inputs, drift paths, termination statuses, and geometry boundaries.
- Measure runtime and memory use.
- Document thread-safety assumptions and build-order requirements.
- Validate field-map and `kEff` settings for supported detector conditions.

AI-generated summaries can contain errors. Verify these points against the implementation and validation results.

### CODE REVIEW

**Comment by pinkenburg** [comment]
The build crashed with an ordering problem, PHGarfield was build after the tpc package. I just changed that, jenkins should be able to build this now

File: offline/packages/PHGarfield/PHGarfield.h [hunk 1]
```diff
@@ -33,7 +33,18 @@ class PHGarfield : public SubsysReco
   int InitRun(PHCompositeNode *) override;
   int process_event(PHCompositeNode *topNode) override;
 
-  bool StopHere(const double x, const double y, const double z, const double zPrevious);
+  enum class ReverseDriftStatus
```
Comment (pinkenburg) [inline offline/packages/PHGarfield/PHGarfield.h:48]: could you add values (Running = 0, CentralMembrane=1,...). That protects against changes by someone adding a member somewhere in the middle.

File: offline/packages/tpc/LaserClusterHelper.cc [hunk 2]
```diff
@@ -39,6 +47,23 @@ void LaserClusterHelper::loadNodes(PHCompositeNode* topNode)
     {
         std::cout << "LaserClusterHelper::loadNodes - TPCGEOMCONTAINER not found on node tree" << std::endl;
     }
+
+    m_phgarfield = std::make_unique<PHGarfield>();
+    m_phgarfield->SetElectricFieldMap("/sphenix/user/dloomis/Distortions/garfield_fields/sphenix_rossegger_garfield_field.root");
+    TVector3 Northxyz(-0.001, -0.001, 1123.109);  // mm
```
Comment (pinkenburg) [inline offline/packages/tpc/LaserClusterHelper.cc]: TVector3 is a legacy class, please replace
https://root.cern/doc/v630/classTVector3.html

File: offline/packages/PHGarfield/PHGarfield.cc [hunk 3]
```diff
@@ -717,15 +729,15 @@ bool PHGarfield::StopHere(const double x, const double y, const double z,
...
     {
-      return true;
+      return ReverseDriftStatus::CentralMembrane;
     }
 
     // Protect against infinite loops if Garfield returns no longitudinal motion.
     if (z == zPrevious)
     {
-      return true;
+      return ReverseDriftStatus::Stuck;
```
[the remark this answers is not in the document]
Comment (devloom) [reply offline/packages/PHGarfield/PHGarfield.cc:1618]: With a step size of 50 ns, the DeltaX and DeltaY between two adjacent points near the central membrane is insignificant, on the order of microns.

[the remark this answers is not in the document]
Comment (devloom) [reply offline/packages/PHGarfield/PHGarfield.cc:1618]: As of now, we don't require z precision for diffuse laser clusters since we do not know the exact start time of the laser flash. Future development could implement an interpolation.

File: offline/packages/tpc/LaserClusterHelper.h [hunk 4]
```diff
@@ -5,33 +5,53 @@
...
 
     bool m_useZ{false};
     bool m_useGlobal{true};
 
+    double m_garfield_cmvoltage{380.0};
+    bool m_garfield_zerofield{false};
+    double m_garfield_spacechargescaleside0{1.0};
+    double m_garfield_spacechargescaleside1{1.0};
+    double m_garfield_stepns{50.0};
+    std::string m_spacechargefieldmap{"/sphenix/user/dloomis/Distortions/garfield_fields/sphenix_rossegger_garfield_field.root"};
```
Comment (osbornjd) [inline offline/packages/tpc/LaserClusterHelper.h]: Is this the same electric field map that Mariia is using in the poly clusterizer? We should have a single map unless there is a good reason not to (I am working on integrating it to pull from the CDB after Chris added it)

Comment (devloom) [reply offline/packages/tpc/LaserClusterHelper.h]: This should be the same map. I can update it to pull from the CDB.

Comment (osbornjd) [reply offline/packages/tpc/LaserClusterHelper.h]: I just merged the changes which shows you how to do this, see https://github.com/sPHENIX-Collaboration/coresoftware/pull/4403/changes

**Comment by devloom** [comment]
This should be ready to be merged
